### DISCUSSION TITLE
Override type resolution for a single type

### DISCUSSION THREAD

**Discussion by rrva** [graphql-java/graphql-java#3013] [Q&A] [answered]
I know this belongs in graphql-java-kickstart, since I am a user of it, but trying here as well since it is related to graphql-java as well.

Is there a way to override the wiring/type resolution for a single type but keep the rest?

I have a GraphQL interface Fruit which has subtypes Apple and Banana. There exists a corresponding java interface Fruit and java classes Apple and Banana .

How do I introduce a new class only on the java side, lets say called Orange, which on the java side implements Fruit, but I want it to resolve to Banana on the GraphQL side (so I do not have to introduce another GraphQL type in the schema).

I have tried using TypeResolvers for this. (See https://github.com/graphql-java-kickstart/graphql-java-tools/discussions/718). 

My problem stems from me wanting to extend an existing GraphQL interface with a new subtype, but this breaks backward compatibility for clients.

I am considering other techniques:

- Introduce a graphQL type as well but implement some solution to keep backwards compatibility by rewriting queries if they perform casts `... on` (or getting rid of casts in queries), and rewriting __typename in responses (could get very messy)

**Comment by btserket** [comment, marked as answer]
From graphql-java's perspective, each individual interface/union types has its own type resolver. So even if Kickstart isn't letting you customize it per-type, you can always transform the result by changing the `GraphQLCodeRegistry` and overriding the specific type resolver.

**Reply by rrva** [reply]
Thanks, I got the type resolver part working, but how do I override the data fetcher for this type? Can this also be done through the registry?

**Comment by bbakerman** [comment]
Assuming that kickstarts type resolves work like the base graphql java ones you would just do this in Java code

```
TypeResolver fruityResolver = trenv -> {
   String typeName = "";
   if (trenv.getObject() instanceof Orange || trenv.getObject() instanceof Banna) {
      typeName = "Banana";
  } else {
      typeName = "Apple"
  }
  return trenv.getSchema().getObjectType(typeName);
}
```

I dont know how kickstart allows you to plugin in `graphql.schema.TypeResolver` instances

**Reply by rrva** [reply]
Yes, that works. With graphql-java-tools I used the following:

```kotlin
val executableSchema = schemaParser.makeExecutableSchema()
val oldFruitTypeResolver =
    executableSchema.codeRegistry.getTypeResolver(GraphQLInterfaceType.newInterface().name("Fruit").build())

val fruitTypeResolver = TypeResolver { env ->
    when {
        env.getObject<Any>() is Orange -> env.schema.getObjectType("Banana")
        else -> oldFruitTypeResolver.getType(env)
    }
}

val schemaObjects = schemaParser.parseSchemaObjects()

return GraphQLSchema.newSchema()
    .description(schemaObjects.description)
    .query(schemaObjects.query)
    .mutation(schemaObjects.mutation)
    .subscription(schemaObjects.subscription)
    .additionalTypes(schemaObjects.dictionary)
    .additionalDirectives(schemaObjects.directives)
    .codeRegistry(

        schemaObjects.codeRegistryBuilder
            .typeResolver("Fruit", fruitTypeResolver)
            .build()
    )
    .build()
```

But, now I need a way to override the data fetchers as well, preferably only for Fruit and the subtypes of Fruit

**Reply by rrva** [reply]
I was able to override the data fetchers using the codeRegistry. The trick was to use the existing schema, and traverse it using a visitor and record all the FieldCoordinates that needed and overridden data fetcher.
